packages/graphql-codegen-cli/src/utils/patterns.ts

```diff
@@ -0,0 +1,308 @@
+  // This means we don't get a nice boolean result and must check array length instead
+  const match = mm([relativeCandidatePath], patterns, { cwd: process.cwd() }).length === 1;
+
+  return match;
+};
+
+/**
+ * Filter the provided list of patterns to include only "affirmative" (non-negated) patterns.
+ *
+ * @param patterns List of micromatch patterns (or paths) to filter
+ */
+const onlyAffirmativePatterns = (patterns: string[]) => {
+  return patterns.filter(pattern => !mm.scan(pattern).negated);
+};
+
+/**
+ * Filter the provided list of patterns to include only negated patterns.
+ *
+ * @param patterns List of micromatch patterns (or paths) to filter
+ */
+const onlyNegatedPatterns = (patterns: string[]) => {
+  return patterns.filter(pattern => mm.scan(pattern).negated);
+};
+
+/**
+ * Given a list of negated patterns, invert them by removing their negation prefix
+ *
+ * If there is a non-negated pattern in the list, throw an error, because this
+ * function should only be called after filtering the list to be only negated patterns
+ *
+ * @param patterns List of negated micromatch patterns
+ */
+const invertNegatedPatterns = (patterns: string[]) => {
+  return patterns.map(pattern => {
+    const scanned = mm.scan(pattern);
+    if (!scanned.negated) {
+      throw new Error(`onlyNegatedPatterns got a non-negated pattern: ${pattern}`);
+    }
+
+    // SAFETY: Defensive programming, it's possible some picomatch/micromatch logic
+    // could call a patter negated but without its prefix !. We don't know how to
+    // handle this, but it's not worth crashing over, so warn the user and leave the pattern untransformed
+    if (scanned.negated && scanned.prefix !== '!') {
+      log(`${logSymbols.warning} Got confusing pattern, appears negated but without ! prefix: ${pattern}`);
+      return pattern;
+    }
+
+    // Remove the leading !
+    return pattern.slice(1);
+  });
+};
+
+/**
+ * Return true if relativeCandidatePath matches any of the affirmativePatterns
+ *
+ * @param relativeCandidatePath A relative path to evaluate against the supplied affirmativePatterns
+ * @param affirmativePatterns A list of patterns, containing no negated patterns, to evaluate
+ */
+const matchesAnyAffirmativePattern = (relativeCandidatePath: string, affirmativePatterns: string[]) => {
+  if (isAbsolute(relativeCandidatePath)) {
+    throw new Error('matchesAny should only be called with relative candidate path');
+  }
+
+  // Developer error: This function is not intended to work with pattern sets including negations
+  if (affirmativePatterns.some(pattern => mm.scan(pattern).negated)) {
+    throw new Error('matchesAnyAffirmativePattern should only include affirmative patterns');
+  }
+
+  // micromatch.isMatch does not omit matches that are negated by negation patterns,
+  // which is why we require this function only examine affirmative patterns
+  return mm.isMatch(relativeCandidatePath, affirmativePatterns);
```

Comment:
Here we _can_ use `micromatch.isMatch`, because we explicitly checked that we're only matching on affirmative patterns. (Recall that `micromatch.isMatch` does not work as expected when patterns include a mix of negated and affirmative patterns.)